tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,555 @@
+        endpoint_provider.resolve_endpoint(**input_params)
+        assert err.msg == expected_object['error']
+
+
+@pytest.mark.parametrize(
+    "ruleset,input_params,expected_object", generated_ruleset_testcases('url')
+)
+def test_endpoint_resolution(
+    partitions, ruleset, input_params, expected_object
+):
+    endpoint_provider = EndpointProvider(ruleset, partitions)
+    endpoint = endpoint_provider.resolve_endpoint(**input_params)
+    assert endpoint.url == expected_object['url']
+    assert endpoint.properties == expected_object.get('properties', {})
+    assert endpoint.headers == expected_object.get('headers', {})
+
+
+@pytest.mark.parametrize(
+    "input_parameters,expected_endpoint",
+    [
+        (
+            {
+                'Region': 'us-east-1',
+                'Bucket': 'arn:aws:s3:us-west-2:123456789012:accesspoint:myendpoint',
+                'Endpoint': 'https://mygreatendpoint.com',
+            },
+            RuleSetEndpoint(
+                url='https://mygreatendpoint.com',
+                properties={
+                    "authSchemes": [
+                        {
+                            "name": "v4",
+                            "signingName": "s3",
+                            "signingScope": "us-east-1",
+                            "uriEscapePath": False,
+                        }
+                    ]
+                },
+                headers={
+                    'x-amz-region-set': [
+                        'us-west-2',
+                    ]
+                },
+            ),
+        ),
+        (
+            {
+                'Region': 'us-east-1',
+                'Bucket': 'arn:aws:s3:us-west-2:123456789012:accesspoint:myendpoint',
+                'Endpoint': 'http://mygreatendpoint.biz.amazonaws.com',
+            },
+            RuleSetEndpoint(
+                url='mygreatendpoint.biz.amazonaws.com',
+                properties={
+                    "authSchemes": [
+                        {
+                            "name": "v4",
+                            "signingName": "s3",
+                            "signingScope": "us-east-1",
+                            "uriEscapePath": False,
+                        }
+                    ]
+                },
+                headers={
+                    'x-amz-region-set': [
+                        'us-east-1',
+                    ]
+                },
+            ),
+        ),
+    ],
+)
+def test_endpoint_resolution_with_is_valid_host_label(
+    partitions, ruleset_dict, input_parameters, expected_endpoint
+):
+    endpoint_provider = EndpointProvider(ruleset_dict, partitions)
+    endpoint = endpoint_provider.resolve_endpoint(**input_parameters)
+    assert endpoint.url == expected_endpoint.url
+    assert endpoint.properties == expected_endpoint.properties
+    assert endpoint.headers == expected_endpoint.headers
+
+
+def test_endpoint_resolution_error_with_is_valid_host_label(
```

Comment:
This may be clearer as `test_endpoint_resolution_invalid_host_label`. It's also not clear how this test is proving the error came from the is_valid_host_label check. Won't any default error satisfy this?

Reply:
agree it's definitely not clear, but it is testing `_is_valid_host_label`. Take a look at the custom rule I wrote `tests/unit/data/endpoints/rules/ruleset.json`